[ICDS Dashboard]: Fixing past data due to changed the height eligiblity

ShubhamBansal1997 opened the pull request:
<!--- Provide a link to the ticket or document which prompted this change -->
https://app.asana.com/0/0/1153564689437802/f
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

### Execution Plan for this Data Correction
- The command takes two inputs
    - first the date from which the execution would be started by default it's `2017-03-01` i.e dashboard initial date
    - second is the script number which is used to specify from which script you should start if you stop the command in between, by default it's 1 i.e. from start

- Now the execution of the scripts
   - There are around 7 scripts

   - This set of scripts are executed in the months containing partitioned tables
      - `fix_past_data_part_1.sql`
      - `fix_past_data_part_2.sql`
      - `fix_past_data_part_3.sql`
      - `fix_past_data_part_4.sql`
      - `fix_past_data_part_5.sql`
      - `fix_past_data_part_6.sql`
      - `fix_past_data_part_7.sql`
   - This set of scripts are executed in the months containing non - partitioned tables
      - `fix_past_data_part_1.sql`
      - `fix_past_data_part_2.sql`
      - `fix_past_data_part_1_3.sql`
      - `fix_past_data_part_1_4.sql`
      - `fix_past_data_part_1_5.sql`
      - `fix_past_data_part_1_6.sql`
      - `fix_past_data_part_1_7.sql` 
 
- Run this `custom/icds_reports/management/commands/fix_past_data.py`

Planned to execute between 6PM to 11 PM IST

##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

ShubhamBansal1997 committed:
[sql]: added the query for fixing the past data which changed due to
height eligibility

- Github Issue:

ShubhamBansal1997 committed:
[fix]: minor typo fixes

- Github Issue:

Rohit25negi commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,93 @@
+         SUM(CASE WHEN chm.nutrition_status_weighed = 1 AND chm.height_measured_in_month = 1 THEN 1 ELSE 0 END) as weighed_and_height_measured_in_month
+
+      FROM "child_health_monthly" chm
+      LEFT OUTER JOIN "awc_location" awc_loc ON (
```
why do you need awc_location here? only awc_id is enough no? which is already there in child_health table

Rohit25negi commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,93 @@
+         ELSE
+            0
+         END as height_measured_in_month
+         FROM child_health_monthly chm
```
you dont need a join here. you can make it same as you updated in your another Pr

Rohit25negi commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,93 @@
+      AND block_id = tmp.block_id
+      AND supervisor_id = tmp.supervisor_id
+      AND awc_id = tmp.awc_id
+      AND month = tmp.month
```
also we need to specify the aggregation level.

Rohit25negi commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,93 @@
+      AND supervisor_id = tmp.supervisor_id
+      AND awc_id = tmp.awc_id
+      AND month = tmp.month
+
```
we would also need to do the rollup to higher levels of location.

Rohit25negi reviewed (changes requested):
Left a few comments

ShubhamBansal1997 committed:
[fix|refactor]
 - refactored the query
 - added cost for the query

- Github Issue:

ShubhamBansal1997 committed:
[refactor]: added the roll up queries

- Github Issue:

ShubhamBansal1997 committed:
[fix]: using the awc_location_local instead of ucr

- Github Issue:

Rohit25negi reviewed (approved):
looks good. @calellowitz  First GM Query look pretty huge to me, do you see any optimization can be done on this?

calellowitz commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,358 @@
+        ELSE
+          current_month_wasting
+        END
+   FROM icds_dashboard_growth_monitoring_forms gm
```
This wont work at all. That table only contains the last 3 months of data.

calellowitz commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,358 @@
+
+-- aggegrate child health table fix
+DROP TABLE IF EXISTS temp_agg_child_my;
+CREATE TABLE temp_agg_child_my AS 
```
This will be much faster if the temp table is distributed

calellowitz commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,358 @@
+WHERE 
+  chm.month = '2018-05-01' 
+GROUP BY 
+  chm.awc_id, 
```
for the distributed table, you will also need to group by supervisor_id

calellowitz commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,358 @@
+    stunting_normal = ut.stunting_normal,
+    height_measured_in_month = ut.height_measured_in_month,
+    weighed_and_height_measured_in_month = ut.weighed_and_height_measured_in_month
+  FROM (
```
I think this can be written far more clear as `FROM temp_agg_child_my AS ut`. The subselect is not necessary

calellowitz commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,358 @@
+        weighed_and_height_measured_in_month
+      ) as weighed_and_height_measured_in_month 
+    FROM 
+      "agg_child_health_2019-01-01_5" agg_child 
```
Why is this month different than the other one?

ShubhamBansal1997 replied:
Updated

calellowitz commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,358 @@
+
+
+UPDATE 
+  "agg_child_health_2019-01-01_3" agg_child_health 
```
Why do the months switch from 2018-05 to 2019-01 halfway through the queries?

calellowitz commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,358 @@
+      ) as weighed_and_height_measured_in_month 
+    FROM 
+      "agg_child_health_2019-01-01_5" agg_child 
+      INNER JOIN (
```
What does this inner join do?

ShubhamBansal1997 replied:
@Rohit25negi Can you answer this one too?

ShubhamBansal1997 committed:
Update fix_past_data_due_to_change_in_height_eligibilty.sql

Rohit25negi commented:
> looks good. @calellowitz First GM Query look pretty huge to me, do you see any optimization can be done on this?

Did not mean to approve it. Was trying to approve https://github.com/dimagi/commcare-hq/pull/26629 and mistakenly approved this.

Rohit25negi reviewed (changes requested):
@ShubhamBansal1997, please address @calellowitz  comments. I assume this is the correction of all historical data. As @calellowitz  mentioned Gm table does not exist so you will have to create it first for all.
for the first oldest month you can use same query used https://github.com/dimagi/commcare-hq/pull/26612. and then use the normal gm aggregation script.

ShubhamBansal1997 committed:
Update fix_past_data_due_to_change_in_height_eligibilty.sql

ShubhamBansal1997 commented:
@calellowitz @Rohit25negi Updated the PR

ShubhamBansal1997 commented:
@calellowitz @Rohit25negi Updated the PR

ShubhamBansal1997 committed:
[build]: added the fill missing gm form script

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[build]: added the management command to execute the past data
correction script

- Github Issue:

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,55 @@
+            }
+            with open(path, "r", encoding='utf-8') as sql_file:
+                sql_to_execute = sql_file.read()
+                sql_to_execute = sql_to_execute % context
```
have you tested that this works the way you expect? I have never seen reading a string like this from a file and then formatting it after the fact. i thought formatting had to happen at string creation

calellowitz replied:
I think what you actually want to do is pass these as params to the sql query

ShubhamBansal1997 replied:
Yeah, it's done to pass the params to the sql query and it works correctly, in fact I am not sure about this initially when I tried in the cas-data-gen, https://github.com/dimagi/cas-data-gen/blob/master/data_gen/data/birth_preparedness.py Here is the sample where I have used the same.

calellowitz replied:
I think it would still be better to pass them as true parameters. using `cursor.execute(query, params)` rather than formatting them ourselves. that lets the underlying library handle the paraterization, and is the more standard way to pass arguments to a sql script

calellowitz replied:
If you are going to do it this way, please use standard string formatting `string.format(...)` rather than the old style `%` formatting

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,55 @@
+
+    def run_task(self, date):
+        initial_date = datetime(2017, 3, 1, 0, 0)
+        intervals = date.month - initial_date.month + 12 * (date.year - initial_date.year) + 1
```
it might be nice to have a comment about what this means. I am getting confused by the date math and the magic numbers

ShubhamBansal1997 replied:
Sure, although it just getting the number of intervals between a set of data, by intervals here we mean the number of month, which I turn used in the next lines to formulate an array of all dates(first day of each month between that internal)

calellowitz commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_1.sql:
```diff
@@ -0,0 +1,33 @@
+        ELSE
+          current_month_wasting
+        END
+   FROM icds_dashboard_growth_monitoring_forms gm
```
This wont work for most of the months (same comment i made earlier)

ShubhamBansal1997 replied:
Added the script to build the past gm forms

calellowitz commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_3.sql:
```diff
@@ -0,0 +1,23 @@
+    stunting_normal = ut.stunting_normal,
+    height_measured_in_month = ut.height_measured_in_month,
+    weighed_and_height_measured_in_month = ut.weighed_and_height_measured_in_month
+  FROM (
```
this would be much clear as `FROM temp_agg_child_my` without the unnecessary subselect

ShubhamBansal1997 replied:
Done

calellowitz commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_2.sql:
```diff
@@ -0,0 +1,51 @@
+DROP TABLE IF EXISTS temp_agg_child_my;
+CREATE TABLE temp_agg_child_my AS 
```
Same comment about this being faster if the temp table is distributed

ShubhamBansal1997 replied:
@Rohit25negi Could you explain this one?

calellowitz replied:
If you distribute the temp table across the citus workers, instead of leaving it as a single local table on the master, the queries that build it will be much quicker

Rohit25negi replied:
May be @calellowitz  can correct me. This table is used to fill int data to agg_Child_health table which is a local table. And since agg_child_health table is local table  it can directly use distributed table to update data in it. So I think this temp table has to be a local table.

Cal m I missing anything about the reason of making this temp table distributed? Because if this temp table is made distributed. it can't be used to update data in agg-child_health.

calellowitz replied:
Yeah, so what we do in the aggregation is to make the temp table distributed for agg level 5, then copy it after its created to the local machine for the updates on the actual table. But it made the overall update time WAY faster to do it that way. You can look at both agg_child_health and agg_awc to see examples of how we do that

Rohit25negi replied:
do you mean to say,
`copy from chm to distributed table + then copying from that table to local table` is quicker than `copying to local table directly like this `
@calellowitz

calellowitz replied:
yes

calellowitz reviewed (changes requested):
I left a few comments but noticed that all of my old comments were marked resolved, but none of the suggestions were addressed. All of the problems are still there and none of the improvements were made, so going to hold off on further review until my initial review is acknowledged

calellowitz commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_3.sql:
```diff
@@ -0,0 +1,23 @@
+UPDATE "agg_child_health_%(start_date)s_5" agg_child_health
```
I don't think this format is valid for all months. At some point last fall we stopped subdividing the monthly agg child health

ShubhamBansal1997 replied:
Updated the script to handle both type of tables

challabeehyv committed:
ui fix for block user tab reports

challabeehyv committed:
ui fix for block user tab reports

ShubhamBansal1997 committed:
Merge branch 'cr/icds-1331' of github.com:challabeehyv/commcare-hq into sb/past_data_correction_height_eligibility

ShubhamBansal1997 committed:
Merge branch 'master' into sb/past_data_correction_height_eligibility

ShubhamBansal1997 committed:
[build]: fix previous scipts and added new ones

- Github Issue:

ShubhamBansal1997 committed:
[fix]: updated the month from which the gm forms starts populating

- Github Issue:

ShubhamBansal1997 committed:
[fix]: updated the files

- Github Issue:

ShubhamBansal1997 committed:
Merge branch 'master' into sb/past_data_correction_height_eligibility

ShubhamBansal1997 committed:
[fix]: abrupt commit

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: abrupt commit issue

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,74 @@
+        initial_3 = datetime(2019, 8, 1, 0, 0)
+        # date at which the partitioning breaks
+        date_breaks = [initial_2, initial_3]
+        # get the number of month intervals between the dates
```
is date going to be the current date?

ShubhamBansal1997 replied:
99 percent it would be current date but still taking as a input from the management command

calellowitz commented on custom/icds_reports/management/commands/fill_missing_gm_forms.py:
```diff
@@ -0,0 +1,38 @@
+                                                    location_type__name='state').values_list('location_id', flat=True))
+        for monthly_date in monthly_dates:
+            for state_id in state_ids:
+                icds_state_aggregation_task(state_id=state_id, date=monthly_date,
```
I dont think this is how you want to do this. Based on the speed that builds, it will take at least 48 hours to build all these tables, while they are building the agg will be slowed down, and as soon as the agg runs it will delete, whatever you have managed to build up to that point in its automatic cleanup. If the issue here is just eligibility, is it possible to use data that is already in `child_health_monthly`? It looks like most of these values are recorded in the table already, without the eligibility checks. You should be able to just copy those if the case meets the new eligibility, rather than reaggregating from scratch, and that would be much more performant. I think it could actually make most of this unnecessary.

ShubhamBansal1997 replied:
I agree to this point to some extent but this field
```sql
height_measured_in_month = CASE
         WHEN (gm.height_child_last_recorded >= '%(start_date)s' AND gm.height_child_last_recorded <'%(end_date)s') AND (valid_in_month=1 AND age_tranche::Integer <= 60) THEN 1
         ELSE
           0
         END,
```
is completely dependent on the gm table

calellowitz replied:
why is that? it appears the height part of that from the gm table is exactly replicated in `recorded_height` in child health monthly. so that expression could be
```
height_measured_in_month = CASE
         WHEN (recorded_recorded is not null) AND (valid_in_month=1 AND age_tranche::Integer <= 60) THEN 1
         ELSE
           0
         END,
```

ShubhamBansal1997 replied:
Updated!!

ShubhamBansal1997 committed:
[fix]: remove the gm script as the field used from the ucr is not
required also removed from the subsequent files too

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[revert]: remove the file as the gm ucr table is not required

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 commented:
@calellowitz @Rohit25negi Updated the changes mentioned

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_1_1.sql:
```diff
@@ -0,0 +1,42 @@
+UPDATE child_health_monthly
```
how is fix_past_data_part_1_1.sql different from fix_past_data_part_1.sql

ShubhamBansal1997 replied:
it was done to make looping thing easy in the script

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_1_4.sql:
```diff
@@ -0,0 +1,78 @@
+UPDATE 
```
you could have combined 1_4 to 1_7 as this is just rollups and filters like aggregation_level etc can be dynamic and passed while invoking. It will reduce the amount of code in this PR

ShubhamBansal1997 replied:
answered below

Rohit25negi commented on custom/icds_reports/utils/fix_past_data_due_to_change_in_height_eligibilty.sql:
```diff
@@ -0,0 +1,358 @@
+-- child health table fix
```
queries in this file looks duplicate to what I see above. Is this file here fore some special reason?

ShubhamBansal1997 replied:
my bad it's the code from some previous similar PR

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_4.sql:
```diff
@@ -0,0 +1,71 @@
+UPDATE 
```
Same comment, I think you can combine file from _4 to _7 into one because the query is mostly same just a few parameters changes

ShubhamBansal1997 replied:
group by parameters are different in all of them

ShubhamBansal1997 replied:
and also the where clause the aggregation level, and planning to execute this weekend so let's keep it that way

Rohit25negi reviewed (changes requested):
generally looks good. Left a few cleanup comments.

ShubhamBansal1997 committed:
[fix]: removed extra file

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[fix]: remove the same code files

- Github Issue:

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,74 @@
+
+    def run_task(self, date):
+        monthly_dates = []
+        initial = datetime(2017, 3, 1, 0, 0)
```
these can be dates not datetimes

ShubhamBansal1997 replied:
keeping it datetimes for calculations although converting them to string dates before moving into scripts

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,74 @@
+                print(sql_to_execute)
+                _run_custom_sql_script(sql_to_execute)
+
+    def handle(self, *args, **kwargs):
```
this allows you to specify and end date but not a start date, which means it is not resumable. This is likely to take a very long time to run, and we will want to do it a few months at a time, so it should be configured to allow picking both a start and end. Alternatively, many of our commands ask you to specify the months directly, which would also make the logic below way simpler.

ShubhamBansal1997 replied:
updated

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,74 @@
+        # date at which the partitioning breaks
+        date_breaks = [initial_2, initial_3]
+        # get the number of month intervals between the dates
+        intervals = date.month - initial.month + 12 * (date.year - initial.year) + 1
```
Why is this all necessary. It looks like it would be way simpler to just take the start date, and increment one month at a time until it is greater than the end date (rather than precalucluating the intervals with this confusing conversation)

ShubhamBansal1997 replied:
So interval things are important the thing, which is due to the partitioned tables, partitioned tables are not continuous, they are present are somewhere and absent at some so need to handle that , I put these two

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,74 @@
+            help='The start date (inclusive). format YYYY-MM-DD'
+        )
+
+    def build_data(self, monthly_date):
```
monthly_date is a confusing name since its not a date, its a dictionary

ShubhamBansal1997 replied:
yeah would change just initially using a date but now changed to dict.

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,74 @@
+
+
+@transaction.atomic
+def _run_custom_sql_script(command, day=None):
```
It looks like day is unused. That can be removed.

ShubhamBansal1997 replied:
updated!

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,74 @@
+            with open(path, "r", encoding='utf-8') as sql_file:
+                sql_to_execute = sql_file.read()
+                sql_to_execute = sql_to_execute % context
+                print(sql_to_execute)
```
what is the print for?

calellowitz replied:
If this is for logging, it would cleaner to just output the query number so its easier to scan the output. Ideally as `running query x of 7 for month Y` so that its clear how many there are to go

ShubhamBansal1997 replied:
yeah logging is there like the way you said and it's pointing out both the number as well as month this on the actual query being logged while running so we can have a glimpse what's the actually query which made the things break

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,74 @@
+            monthly_dates.append({"record_date": initial, "default": default})
+            initial = initial + relativedelta(months=1)
+
+        for monthly_date in monthly_dates:
```
It would be good to add logging here so we can keep track of the progress, like outputting what date we are on

ShubhamBansal1997 replied:
it's already there in the above function

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,74 @@
+            print(f'==============Executing Script {i} =============')
+            if monthly_date["default"] is False and i > 2:
+                path = os.path.join(os.path.dirname(__file__), 'sql_scripts',
+                                'fix_past_data_part_1_{}.sql'.format(i))
```
this naming scheme is really confusing, why does part 1_3 come after part 2? it would be easier to understand if instead of a number, that was a descriptive word like "partitioned" so that it was clear what the difference was and for all cases the ordering is clear (1 then 2 then 3 etc) rather than (1 then 2 then 1_3)

ShubhamBansal1997 replied:
so naming scheme is simple basically for one set we are running `_1_{}` and for the other `_{}` since the first two queries are same so we are running same for them that's why using the >2 thing.

calellowitz commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_2.sql:
```diff
@@ -0,0 +1,74 @@
+  coalesce_disabled, 
+  coalesce_minority, 
+  coalesce_resident 
+ORDER BY 
```
what does this order by do?

ShubhamBansal1997 replied:
updated

calellowitz reviewed (changes requested):
Can you explain a bit more how you plan to run this. I think we will want to run this like most data corrections in small increments during off peak times and when the agg is not running, and that does not seem possible as it is currently written

ShubhamBansal1997 committed:
[fix]: fixes to the script [ci skip]

- Github Issue:

ShubhamBansal1997 committed:
[fix]: updated script as per reviews and added resume capablity too

ShubhamBansal1997 commented:
@calellowitz Updated as per all the comments

ShubhamBansal1997 committed:
[fix ]: missing sql command

ShubhamBansal1997 committed:
[revert]: remove extra files

ShubhamBansal1997 committed:
[revert]: extra change

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
[revert]: reverting

ShubhamBansal1997 committed:
Merge branch 'master' into sb/past_data_correction_height_eligibility

ShubhamBansal1997 committed:
[rvert]: .gitmodules file

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_1.sql:
```diff
@@ -0,0 +1,43 @@
+        ELSE
+          0
+        END,
+      current_month_stunting = CASE
```
I think calculation would be wrong  for kids who are <6 months because for previous tables these columns  were already not having values for children <6 months.  Eg. In Jan 2019, for kid of 3 months, current_month_stunting was already 0/null because of height eligibility condition applied to it was 6<eligible<=60. Or you are saying height eligibility was not applied to them before?
If no then these calculations needs to updated. 
Example: current_month_stunting is prepared by `zscore_grading_hfa_last_recorded` and `zscore_grading_hfa` which are already stored in the chm table without height eligibility check. Those should be used here.

This is same for all columns below

ShubhamBansal1997 replied:
Updated!!

Rohit25negi commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,82 @@
+
+        for monthly_date_dict in monthly_dates_list:
+            self.build_data(monthly_date_dict, script_number)
+            script_number = 1
```
Nit: if this is always 1. why to create a variable.

ShubhamBansal1997 replied:
It's no always 1, if we break the script in between it will start from the breaked script number

Rohit25negi reviewed (changes requested):
Left a few comments. Only one is blocking.

ShubhamBansal1997 committed:
[fix]: review fixes

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_1_3.sql:
```diff
@@ -0,0 +1,18 @@
+    agg_child_health.awc_id=ut.awc_id and 
+    agg_child_health.month=ut.month and
+    agg_child_health.age_tranche=ut.age_tranche and
+    agg_child_health.caste=ut.caste and
```
The agg level 5 group is wrong here. it also include other fields as well https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/utils/aggregation_helpers/distributed/agg_child_health.py#L185

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_1_3.sql:
```diff
@@ -0,0 +1,18 @@
+    agg_child_health.month=ut.month and
+    agg_child_health.age_tranche=ut.age_tranche and
+    agg_child_health.caste=ut.caste and
+    agg_child_health.aggregation_level = 5
```
The agg level 5 group is wrong here. it also include other fields as well https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/utils/aggregation_helpers/distributed/agg_child_health.py#L185

ShubhamBansal1997 replied:
updated!

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_2.sql:
```diff
@@ -0,0 +1,70 @@
+      chm.awc_id,
+      chm.supervisor_id,
+      chm.month,
+      chm.age_tranche,
```
The agg level 5 group is wrong here. it also include other fields as well https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/utils/aggregation_helpers/distributed/agg_child_health.py#L185

ShubhamBansal1997 replied:
updated!

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_3.sql:
```diff
@@ -0,0 +1,17 @@
+    agg_child_health.awc_id=ut.awc_id and 
+    agg_child_health.month=ut.month and
+    agg_child_health.age_tranche=ut.age_tranche and
+    agg_child_health.caste=ut.caste
```
same group by comment as above

ShubhamBansal1997 replied:
updated!

Rohit25negi reviewed (changes requested):
hey. I think the group at agg_level=5 needs more parameters. Otherwise its fine.
Also good to remove the submodule from the change diff.

ShubhamBansal1997 committed:
[fixed]: fixed queries [skip ci]

Rohit25negi commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_3.sql:
```diff
@@ -0,0 +1,21 @@
+    agg_child_health.caste=ut.caste and
+    agg_child_health.disabled=ut.disabled and
+    agg_child_health.minority=ut.minority and
+    agg_child_health.resident=ut.resident
```
same column name comment as above

ShubhamBansal1997 replied:
same as above

Rohit25negi reviewed (changes requested):
hey, can you try running an entire set of queries locally via the command?

ShubhamBansal1997 commented on custom/icds_reports/management/commands/sql_scripts/fix_past_data_part_1_3.sql:
```diff
@@ -0,0 +1,22 @@
+    agg_child_health.gender=ut.gender and
+    agg_child_health.age_tranche=ut.age_tranche and
+    agg_child_health.caste=ut.caste and
+    agg_child_health.disabled=ut.disabled and
```
actually I have created temp table via agg_child_health, there we are using `minority` not `coalesce_minority`

Rohit25negi reviewed (approved):
Looks good to me. I will let @calellowitz  see if this looks good to him as well as he had already given good feedback to this PR.

ShubhamBansal1997 committed:
[fix]: submodule fix

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
Merge branch 'master' into sb/past_data_correction_height_eligibility

ShubhamBansal1997 commented:
@calellowitz Please review

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,82 @@
+        # default is used to differentiate between the partitioned table monthly date
+        # false -> month got a partitioned table
+        # true -> month got a non partitioned table
+        default = False
```
it would be clearer to name the variable `partitioned` (or the opposite in this case) rather than have to explain it via comment, although the comment is appreciated

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,82 @@
+            date_itr = date_itr + relativedelta(months=1)
+
+        for monthly_date_dict in monthly_dates_list:
+            self.build_data(monthly_date_dict, script_number)
```
i think you can call this directly in the loop above, rather than loop over all the dates to build the dict, then loop over them again to call

calellowitz commented on custom/icds_reports/management/commands/fix_past_data.py:
```diff
@@ -0,0 +1,82 @@
+            print(f'==============Executing Script {i} =============')
+            if monthly_date_dict["default"] is False and i > 2:
+                path = os.path.join(os.path.dirname(__file__), 'sql_scripts',
+                                    'fix_past_data_part_1_{}.sql'.format(i))
```
I still find this name scheme confusing

calellowitz reviewed (changes requested):
If you are planning to merge this, you should the submodule update, otherwise, I am deferring to Rohit on the SQL. The python in the command is looking a lot better and I dont want to block this on anything else, so will let you view my comments as merely suggestions. Thanks for following through on this

calellowitz reviewed (commented):
The other review should have been comment, since its approved to run, but not merge

ShubhamBansal1997 committed:
[fix]: changed end date to the jan
break the second query file into parts to execute one query at a time

- Github Issue:

Authored-by: Shubham Bansal <shubhambansal17@hotmail.com>
Signed-off-by: Shubham Bansal <shubhambansal17@hotmail.com>

ShubhamBansal1997 committed:
:bug: added the missing column
